### PR TITLE
DEVX-2488: TIMESTAMP datatype, timezone conversion, JDBC sink connector KT

### PR DESCRIPTION
**Pull request by awalther28**
### Description

https://confluentinc.atlassian.net/browse/DEVX-2488

Adds a new KT for ksqlDB 0.16.0/0.17. The release is TBD.

Please contact me if you can't pull the RC image, I have been using a local tar instead of pulling from AWS ECR. 
To pull image use the following:
```
aws ecr get-login-password | docker login --username AWS --password-stdin <dev-aws-account-number>.dkr.ecr.us-west-2.amazonaws.com

docker image pull <dev-aws-account-number>.dkr.ecr.us-west-2.amazonaws.com/confluentinc/ksqldb-server:0.16.0-rc891
```
Thanks in advance. 

### TODO Before Merge

- [x] Update ksqldb-server to 0.17.0
- [x] Update ksqldb-cli to 0.17.0
- [x] Validate KT semaphore passes
- [x] Rename title (remove `[DO NOT MERGE]`)
- [ ] (post merge) Do a full KT release

### Staging Docs

http://kafka-tutorials-staging.s3-website-us-west-2.amazonaws.com/DEVX-2488/
http://kafka-tutorials-staging.s3-website-us-west-2.amazonaws.com/DEVX-2488/connect-sink-timestamp/ksql.html

### CODE REVIEW

File: index.html
Comment (ybyzek) [inline index.html:164]: 1. Move below `Add a key to a stream`?
2. To keep the titles short, WDYT:

```suggestion
            <li><a href="connect-sink-timestamp/ksql.html">Convert timezone before sending to a database</a></li>
```

File: _data/tutorials.yaml
Comment (ybyzek) [inline _data/tutorials.yaml:680]:
```suggestion
  introduction: "Consider the requirement to create reports from a database, for which all of the timestamps must be in a timezone which happens to be different from the timezone of the Kafka data source. This examples shows how you can convert timestamp data into another timezone and use a Kafka connector to send that data to Postgres database."
```

File: _data/tutorials.yaml
Comment (ybyzek) [inline _data/tutorials.yaml:676]:
```suggestion
  title: "Timezone conversion and Kafka Connect JDBC sink"
```

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/answer/short-answer.adoc
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/answer/short-answer.adoc:1]:
```suggestion
First use the ksqlDB function `FROM_UNIXTIME` to convert a data field type from `BIGINT` to `TIMESTAMP`, and use the ksqlDB function `CONVERT_TZ` to change the timestamp to the desired timezone.
```

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/answer/short-answer.adoc
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/answer/short-answer.adoc:7]:
```suggestion
Then create a sink connector that reads from the ksqlDB output topic with the converted timezone data.
```

File: _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml:87]: Users may not understand that this tutorial is using ksqlDB's embedded connect worker or that they need to add the connector jars using `confluent-hub install`.  Please either:

- explain these 2 things in the tutorial...or...
- convert to using a standalone connector (and use Dockerfile)

File: _includes/tutorials/connect-sink-timestamp/ksql/code/tutorial-steps/dev/create-stream.sql
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/code/tutorial-steps/dev/create-stream.sql:1]: Presumably it's for demo purposes, but FYC explain why eventTime is set to `BIGINT` at stream creation

File: _includes/tutorials/connect-sink-timestamp/ksql/code/tutorial-steps/dev/select-from-stream.sql
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/code/tutorial-steps/dev/select-from-stream.sql:1]: Before showing it as Denver timezone, would it be helpful to first show the following (without timezone change) to drive home the point that these are 2 distinct steps/functions?

```
SELECT TEMPERATURE, FROM_UNIXTIME(EVENTTIME) AS
....
```

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc:1]: "it is" -- what is it?

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/make-docker-compose.adoc
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/make-docker-compose.adoc:1]: This docker-compose doesn't just obtain Confluent Platform, it also spins up the postgres database. I think it would be helpful to either make postgres bringup a discrete step, or expand the text to tell the reader the full extent of this docker compose.

File: _data/harnesses/connect-sink-timestamp/ksql.yml
Comment (ybyzek) [inline _data/harnesses/connect-sink-timestamp/ksql.yml:84]: Not sure how consistent the tutorials are on this point, but WDYT about adding a Cleanup step, e.g. http://kafka-tutorials-staging.s3-website-us-west-2.amazonaws.com/DEVX-2488/connect-add-key-to-source/ksql.html#clean-up

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/describe-raw-stream.adoc
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/describe-raw-stream.adoc:13]:
```suggestion
Notice that `EVENTTIME` is of type `BIGINT`, this is important to note. In order to convert this column from `UTC` to `America/Denver` time, the column needs to be a `TIMESTAMP` data type. Both `BIGINT` (time in ms) or string representation of a formatted timestamp (`yyyy-MM-dd HH:mm:ss.SSS`) can be converted to a `TIMESTAMP` data type.
```

**Review by ybyzek — COMMENTED** [review]
@awalther28 thanks for the PR!  I made a first pass with comments from reading the staging doc.  I did not run this yet.

File: _data/tutorials.yaml
Comment (awalther28) [reply _data/tutorials.yaml:676]: Did you know there are technically 3 correct ways to use the word time zone. This is technically a correct usage but not when directed towards Americans :) I think it should be changed to match the usage in the rest of the tutorial

Comment (ybyzek) [reply _data/tutorials.yaml:676]: TIL!  Agree, let's just be consistent throughout the tutorial ;)

File: _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml:95]: Should this version match the server version?

File: _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml:47]: This is not consistent throughout ALL the tutorials, but most of them use `ksqldb-server` instead of `ksqldb`.  Would it make sense to use `ksqldb-server` for `hostname` and `container_name` here, as well will match line 45?

File: index.html
Comment (awalther28) [reply index.html:164]: My leaning would be to put it on the stop simply because it's new, but I don't feel strongly about that :) @ybyzek why add it below "add a key to a stream"?

File: _data/harnesses/connect-sink-timestamp/ksql.yml
Comment (awalther28) [reply _data/harnesses/connect-sink-timestamp/ksql.yml:88]: I think that's a clean way to end things. I added it in my last commit d22ac26

File: _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml:87]: I do like the embedded connect worker, it's what the ksql connector uses in connect-add-key-to-source KT. I added a small blurb about this when creating the docker-compose.yml file (https://github.com/confluentinc/kafka-tutorials/blob/DEVX-2488/_includes/tutorials/connect-sink-timestamp/ksql/markup/dev/make-docker-compose.adoc) Not sure if I need to go into more detail. I was sorta following the level of explanation used in the other ksql connector KT.

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/make-docker-compose.adoc
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/make-docker-compose.adoc]: That is a good point, a lot of "stuff" happens in this docker-compose. Similar to a comment above, I did address this at a high level. (https://github.com/confluentinc/kafka-tutorials/pull/760#discussion_r595541683) I can go into more detail if you'd like.

File: index.html
Comment (ybyzek) [reply index.html]: My leaning is that we should promote new KTs on Confluent Developer, and on the KT front page we order the KTs logically.  So adding it below `add a key to a stream` seemed logical to show source before sink.  WDYT?

File: _includes/tutorials/connect-sink-timestamp/ksql/code/tutorial-steps/dev/create-stream.sql
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/code/tutorial-steps/dev/create-stream.sql]: I addressed this in one of my latest commits. After creating the stream I mention:
"The schema for our example stream specifies `eventTime` as a `BIGINT`. It is common for applications to send data in Unix time, thus why `eventTime` is specified as a `BIGINT`."

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc]: addressed in the latest commit https://github.com/confluentinc/kafka-tutorials/pull/760/commits/fe37cdbb1eb4538e6474ccb6e0f8f38e283efa77

File: _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml]: adjusted accordingly in latest commits

File: _data/harnesses/connect-sink-timestamp/ksql.yml
Comment (ybyzek) [inline _data/harnesses/connect-sink-timestamp/ksql.yml:18]: Should we expand this title as well, since it's more than just Confluent Platform?

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc:1]: WDYT about using the word `deploy` instead of `create`?  Goal being to differentiate jar installation as done earlier with `confluent-hub install`.

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc:7]: The information in the sentence is awesome, but the sentence feels misplaced/orphaned.  Any way to tie it in a bit more?  Either move it to another section, or keep it here and add a sentence (like...not verbatim...if you're going the SMT route, you'd add the SMT in this connector config above)

**Review by ybyzek — APPROVED** [review]
👍 @awalther28 nice PR in time for the new ksqlDB release.  I left a few more comments, and I know the error at the end is being fixed up.  So as long as the test completes, LGTM and approved.  THANK YOU !

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/make-docker-compose.adoc
Comment (ybyzek) [reply _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/make-docker-compose.adoc]: LGMT

File: _includes/tutorials/connect-sink-timestamp/ksql/code/tutorial-steps/dev/select-from-stream.sql
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/code/tutorial-steps/dev/select-from-stream.sql]: I broke out this tutorial into more steps so that it's not one nested query. I did keep the nested query as the "short answer"

File: _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/code/docker-compose.yml]: I updated it to 0.15.0 and things appears to work just fine. I will update this to the latest 0.16.0 once that is publically available. Will update the description of the PR to remind myself to do that prior to merge.

File: _data/harnesses/connect-sink-timestamp/ksql.yml
Comment (awalther28) [reply _data/harnesses/connect-sink-timestamp/ksql.yml]: I made it more general, "Set up environment"

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc]: changed

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc:7]: hmm, thinking about where this should live.... I think near the top as part would make the most sense. Will comment once moved

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/answer/short-answer.adoc
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/answer/short-answer.adoc:7]:
```suggestion
Afterward, deploy a sink connector that reads from the ksqlDB output topic with the converted timezone data and writes to the target database.
```

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc:7]: Screen Shot 2021-03-17 at 8 44 02 AM (2)

@ybyzek what are your thoughts on above?

Comment (ybyzek) [reply _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc:7]: @awalther28 the callout is super cool 🥇   There is another drawback to SMTs -- they are not supported in Confluent Cloud -- but the KT should not draw attention to that ;) I do have minor edit suggestions, will drop it in the PR momentarily

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc:10]:
```suggestion
If you were going this route of using `TimestampConverter`, you would add the SMT into the connector configuration, something similar to the following:
```

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc:19]: @awalther28 I brought up the "drawbacks" into the paragraph above to keep the text together, and to shorten it.  If you agree with this change, please delete these 3 lines here, and of course wordsmith the text above as you see fit.

Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc]: :ack: I removed those 3 lines in my latest commit and verified the KT site builds locally.

**Comment by ybyzek** [comment]
@awalther28 outstanding PR, my approval still stands (assuming tests pass).  Also cc: @bbejeck not sure if this PR needs a second set of eyes or not.

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/select-from-stream-unix.adoc
Comment (bbejeck) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/select-from-stream-unix.adoc:1]: Minor point - I found it a little confusing that `CONVERT_TZ` is described and referenced a couple of times, but then the text goes on to use `FROM_UNIX 

Maybe say something along the lines of "Our first step is to convert the `BIGINT` into a `TIMESTAMP` " and move everything about `CONVERT_TZ` to the next section where the tutorial uses the `CONVERT_TZ` function.

But this could stay as is, I'll leave it to you.

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (bbejeck) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc:1]: For your consideration: maybe mention that the sink connector writes to the database.  I know most readers would get this, but a brand new user could miss something.

**Review by bbejeck — APPROVED** [review]
Good tutorial @awalther28 I have a couple of optional comments, but otherwise LGTM

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc]: Good point. I added a comment to address this in my latest commit. https://github.com/confluentinc/kafka-tutorials/pull/760/commits/9904d2ac9ffb1d0894c431caf88c786ba19ad091

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/select-from-stream-unix.adoc
Comment (awalther28) [reply _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/select-from-stream-unix.adoc]: I can see how that is confusing. I added a comment to address this in my latest commit. https://github.com/confluentinc/kafka-tutorials/pull/760/commits/9904d2ac9ffb1d0894c431caf88c786ba19ad091

**Comment by awalther28** [comment]
I believe all of the comments have been addressed @ybyzek and @bbejeck thank you for reviewing and giving your approvals! In commit https://github.com/confluentinc/kafka-tutorials/pull/760/commits/97c3e5c140787e2bd03d050507d4decc2fe5b360 I updated the ksqlDB server and cli version to the actual release version (which has not yet been dropped). This PR is ready to merge.

File: _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc
Comment (ybyzek) [inline _includes/tutorials/connect-sink-timestamp/ksql/markup/dev/create-sink-connector.adoc:9]: @awalther28 should this be:

```suggestion
Prior to version 0.17.0, ksqlDB did not have a `TIMESTAMP` data type so the only way to convert `BIGINT` to a `TIMESTAMP` was with Kafka Connect's Single Message Transforms (SMT), specifically the  https://docs.confluent.io/platform/current/connect/transforms/timestampconverter.html[TimestampConverter]; .
```

**Review by ybyzek — COMMENTED** [review]
Note minor suggestion before merge.  Interesting situation because technically 0.16 is correct but 0.16 wasn't released.  Up to you whether to change.

**Pull request merged**
